scripts/delete-federation.sh

```diff
@@ -28,41 +28,45 @@ NS=federation-system
 
 PUBLIC_NS=kube-multicluster-public
 
-# Remove the federation service account for the current context.
-CONTEXT="$(kubectl config current-context)"
-SA_NAME="${CONTEXT}-${CONTEXT}"
-${KCD} -n ${NS} sa "${SA_NAME}"
-${KCD} -n ${NS} clusterrole "${NS}-controller-manager:${SA_NAME}"
-${KCD} -n ${NS} clusterrolebinding "${NS}-controller-manager:${SA_NAME}"
-
-CONTEXT="$(kubectl config current-context)"
-
-# Remove permissive rolebinding that allows federation controllers to run.
-${KCD} clusterrolebinding federation-admin
+# Remove public namespace
+${KCD} namespace ${PUBLIC_NS}
 
-# Remove role and role binding added by kubefed
-CONTROLLER_ROLE="federation-controller-manager:${CONTEXT}-${CONTEXT}"
-${KCD} clusterrolebinding "${CONTROLLER_ROLE}"
-${KCD} clusterrole "${CONTROLLER_ROLE}"
+# Remove cluster registry CRD
+${KCD} -f vendor/k8s.io/cluster-registry/cluster-registry-crd.yaml
 
-# Delete federated cluster
-# TODO(marun) Remove when federated clusters are stored in federation namespace
-if kubectl get federatedcluster &>/dev/null; then
-  ${KCD} federatedcluster "${CONTEXT}"
-fi
+# Unjoin clusters by removing objects added by kubefed.
+HOST_CLUSTER="$(kubectl config current-context)"
+JOIN_CLUSTERS="${HOST_CLUSTER} ${@}"
+for c in ${JOIN_CLUSTERS}; do
+  SA_NAME="${c}-${HOST_CLUSTER}"
+  CONTROLLER_ROLE="federation-controller-manager:${SA_NAME}"
 
-${KCD} -f hack/install-latest.yaml
-${KCD} -f vendor/k8s.io/cluster-registry/cluster-registry-crd.yaml
+  if [[ "${c}" != "${HOST_CLUSTER}" ]]; then
+    ${KCD} ns ${NS} --context=${c}
+  else
+    ${KCD} sa ${SA_NAME} --context=${c} -n ${NS}
+  fi
 
-# Remove public namespace
-${KCD} namespace ${PUBLIC_NS}
+  ${KCD} clusterrolebinding ${CONTROLLER_ROLE} --context=${c}
+  ${KCD} clusterrole ${CONTROLLER_ROLE} --context=${c}
+  ${KCD} clusters ${c} -n ${PUBLIC_NS}
+  ${KCD} federatedclusters ${c} -n ${NS}
+  CLUSTER_SECRET=$(kubectl -n ${NS} get secrets \
+    -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}' | grep "${c}-*")
+  ${KCD} secret ${CLUSTER_SECRET} -n ${NS}
+done
 
-# Remove crds
-# Enable available types
+# Disable available types
 for filename in ./config/federatedtypes/*.yaml; do
   ${KCD} -f "${filename}"
 done
 
+# Remove federation CRDs, namespace, RBAC and deployment resources.
+${KCD} -f hack/install-latest.yaml
```

Comment:
Should this be configurable to allow targeting `hack/install.yaml` (for developers)?

Reply:
Perhaps. Although I think we should use the same committed config for developers and users.

Reply:
Regarding developers, `hack/install.yaml` is the path that kubebuilder generates to, and so long as it is not idempotent (the order of groups changes with every invocation) I don't think that file should ever be committed.  For users, I don't think they should be using the tree at all.  Release artifacts should be available that allow deployment and use of federation without ever checking out the source tree.

Reply:
Yes, I noticed the RBAC `federation-role` `ClusterRole` config changes the ordering of the `apiGroups` field. That's a bug that should be fixed in `kubebuilder`. I think that would allow us to commit `install.yaml`. Otherwise, we should make sure our docs are consistent with our scripts to only ever use 1 file for our releases and developer workflows, wherever we decide to commit it.

While I don't think users should have to rely on checking out the tree, they should use the tree to apply whatever release artifacts they want. That is, every released version of the config should be available via github at that specific tagged point in the tree e.g. `https://raw.githubusercontent.com/kubernetes-sigs/federation-v2/<release-version-tag>/path/to/install.yaml` and the image should point to that tagged/released version. I think this would allow us to use 1 config file for both workflows.